File: packages/input/src/latency.ts
type TimestampEntry = {
  stage: string;
  time: Date;
};

export class LatencyTracker {
  sequence_id: string;
  timestamps: TimestampEntry[];
  metadata?: Record<string, any>;

  constructor(sequence_id: string, timestamps: TimestampEntry[] = [], metadata: Record<string, any> = {}) {
    this.sequence_id = sequence_id;
    this.timestamps = timestamps;
    this.metadata = metadata;
  }

  addTimestamp(stage: string): void {
    const timestamp: TimestampEntry = {
      stage,
      time: new Date(),
    };
    this.timestamps.push(timestamp);
  }

  // Calculates the total time between the first and last recorded timestamps.
  getTotalLatency(): number {
    if (this.timestamps.length < 2) return 0;

    const times = this.timestamps.map((entry) => entry.time.getTime());
    const minTime = Math.min(...times);
    const maxTime = Math.max(...times);
    return maxTime - minTime;
  }

  toJSON(): Record<string, any> {
    return {
      sequence_id: this.sequence_id,
      timestamps: this.timestamps.map((entry) => ({
        stage: entry.stage,
        // Fill nanoseconds with zeros to match the expected format
        time: entry.time.toISOString().replace(/\.(\d+)Z$/, ".$1000000Z"),
      })),
      metadata: this.metadata,
    };
  }

  static fromJSON(json: any): LatencyTracker {
    const timestamps: TimestampEntry[] = json.timestamps.map((ts: any) => ({
      stage: ts.stage,
      time: new Date(ts.time),
    }));
    return new LatencyTracker(json.sequence_id, timestamps, json.metadata);
  }
}

File: packages/server/src/latency.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct TimestampEntry {
    pub stage: String,
    pub time: String, // ISO 8601 timestamp
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LatencyTracker {
    pub sequence_id: String,
    pub timestamps: Vec<TimestampEntry>,
    pub metadata: Option<HashMap<String, String>>, // Optional metadata
}

impl LatencyTracker {
    // Creates a new LatencyTracker
    pub fn new(sequence_id: String) -> Self {
        Self {
            sequence_id,
            timestamps: Vec::new(),
            metadata: None,
        }
    }

    // Returns the sequence ID
    pub fn sequence_id(&self) -> &str {
        &self.sequence_id
    }

    // Adds a timestamp for a specific stage
    pub fn add_timestamp(&mut self, stage: &str) {
        // Ensure extremely precise UTC format (YYYY-MM-DDTHH:MM:SS.658548387Z)
        let now = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Nanos, true);
        self.timestamps.push(TimestampEntry {
            stage: stage.to_string(),
            time: now,
        });
    }

    // Calculate total latency (first to last timestamp)
    pub fn total_latency(&self) -> Option<i64> {
        if self.timestamps.len() < 2 {
            return None;
        }
        let parsed_times: Result<Vec<_>, _> = self
            .timestamps
            .iter()
            .map(|entry| chrono::DateTime::parse_from_rfc3339(&entry.time))
            .collect();
        if let Ok(parsed_times) = parsed_times {
            let min_time = parsed_times.iter().min().unwrap();
            let max_time = parsed_times.iter().max().unwrap();
            Some((*max_time - *min_time).num_milliseconds())
        } else {
            None
        }
    }
}
